# Communicating with servers via HTTP

## `requests`

`requests` is a third-party package that faciliates the task of sending HTTP requests to communicate with servers.

- To install `requests` with `conda` run:

```bash
conda install requests
```

- To install `requests` with `pip` run:
```bash
pip install requests
```

You can try installing packages using the **Terminal** (Mac), **Anaconda Prompt** (Windows), or by launching a **Terminal** from **Jupyter notebook** (`New` => `Terminal` from the Jupyter notebook `Files` interface).

Whilst `requests` can be used to retrieve HTML from web pages, it shines when it comes to communicating with web APIs that require several arguments.

Let's first import `requests` and retrieve a web page.

We can call `get()` with an URL and save the `Response` object.

In [402]:
import requests

# Get the response from a page
response = requests.get("https://cambridgespark.com")

Let's check the status code, `200` means `OK` -- the request was successful.

In [403]:
response.status_code

200

We can also access the content of the page with the attribute `.content`

In [404]:
# response.content
type(response.content)

bytes

In [405]:
# practice scraping links from page

from bs4 import BeautifulSoup

soup = BeautifulSoup(response.text)
# print(soup.prettify())

# get page links 
links = soup.find_all('a')
links = [link.get('href') for link in links]
if not links:
    print('found no tables')
if links:
    print(f'found {len(links)} links')

# get table data
tables = soup.findChildren('th')
if not tables:
    print('found no tables')


found 65 links
found no tables


Let's look at how `requests` helps when we work with more complex web APIs.

We've created our own web API, you can access it here [here](https://europe-west2-kate-dev.cloudfunctions.net/banking)

We define below the base URL for the API so you do not have to type the whole URL everytime and can just append the endpoint you want to call.

In [406]:
BASE_URL = "https://europe-west2-kate-dev.cloudfunctions.net/banking"

Have a look at the [API Documentation](https://europe-west2-kate-dev.cloudfunctions.net/banking/documentation) to know more about what functionalities are available.

To start with, we will try to list all available users.

To do so, we can call the `/api/users` endpoint.

Let's call our first endpoint to get all available users:

In [407]:
url_endpoint = f"{BASE_URL}/api/users"

r = requests.get(url=url_endpoint)

In [408]:
# Let's verify we got a status code 200 first
r.status_code

401

Oops, error [401](https://httpstatuses.com/401) stands for `Unauthorized` - HTTP is a well defined protocol where each error corresponds to a specific issue. 

We can check the content returned by the API for more details:

In [409]:
r.content

b'Invalid token.'

We "forgot" to mention that our API requires authentication... (see documentation for more details on how to authenticate)

Users need to pass an API key in the headers to authenticate, you can see it as a password. We provide the API key below.

Note: with actual APIs, the documentation will explain how to pass a key to authenticate, it often follows a similar process we are using here.

Thanksfully `requests` allows us to easily define the headers we want to pass when making a request:

In [410]:
api_key = {"Authorization": "NRCqpfD3"}

In [411]:
# save for safety
import json

json.dumps(api_key)

with open('data/api_key.json', 'w') as file:
    json.dump(api_key, file)
    


Let's try again, with our API key this time

In [412]:
url_endpoint = f"{BASE_URL}/api/users"

r = requests.get(url=url_endpoint, headers=api_key)

r.status_code

200

Status code is [200](https://httpstatuses.com/200), great!

Let's see the content:

In [413]:
r.content

b'{"1":"Caroline","2":"Marium","3":"John","4":"John","5":"John","6":"Vic_Flores","7":"Payton","8":"Jess","9":"Sam","10":"Luning","11":"Karl","12":"Johan","13":"username","14":"test","15":"Bethan","16":"Linying","17":"Linying","18":"Vic_Flores","19":"\\ud83d\\ude00","20":"\\");DROP TABLE * ;","21":"D&G","22":"Ben S","23":"Ben S","24":"Michael","25":"John","26":"Emma","27":"Emma","28":"Jasmin","29":"Leo","30":"Matt","31":"Rasool","32":"John","33":"John","34":"John","35":"Amy","36":"Jeremy","37":"Xiong","38":"John","39":"John","40":"John","41":"John","42":"John","43":"Jeremy","44":"John","45":"Maxi Madcock","46":"Tilly Tats","47":"Raine The Paine","48":"John","49":"John","50":"John","51":"John","52":"Jon","53":"Zak","54":"John","55":"Karl","56":"Ben","57":"Karl 2","58":"Karl Hampton","59":"Abby","60":"Kristi","61":"Abby","62":"Benjo","63":"JonB","64":"ZAK","65":"Samantha","66":"John","67":"John","68":"Oli","69":"Oli","70":"Barry","71":"Jennie","72":"Rik","73":"Fahis","74":"Bobandy","75":"

Since our API return JSON objects, we can now use `.json()` directly from `requests` to load the data as a dictionary.

In [414]:
users = r.json()
users

{'1': 'Caroline',
 '2': 'Marium',
 '3': 'John',
 '4': 'John',
 '5': 'John',
 '6': 'Vic_Flores',
 '7': 'Payton',
 '8': 'Jess',
 '9': 'Sam',
 '10': 'Luning',
 '11': 'Karl',
 '12': 'Johan',
 '13': 'username',
 '14': 'test',
 '15': 'Bethan',
 '16': 'Linying',
 '17': 'Linying',
 '18': 'Vic_Flores',
 '19': '😀',
 '20': '");DROP TABLE * ;',
 '21': 'D&G',
 '22': 'Ben S',
 '23': 'Ben S',
 '24': 'Michael',
 '25': 'John',
 '26': 'Emma',
 '27': 'Emma',
 '28': 'Jasmin',
 '29': 'Leo',
 '30': 'Matt',
 '31': 'Rasool',
 '32': 'John',
 '33': 'John',
 '34': 'John',
 '35': 'Amy',
 '36': 'Jeremy',
 '37': 'Xiong',
 '38': 'John',
 '39': 'John',
 '40': 'John',
 '41': 'John',
 '42': 'John',
 '43': 'Jeremy',
 '44': 'John',
 '45': 'Maxi Madcock',
 '46': 'Tilly Tats',
 '47': 'Raine The Paine',
 '48': 'John',
 '49': 'John',
 '50': 'John',
 '51': 'John',
 '52': 'Jon',
 '53': 'Zak',
 '54': 'John',
 '55': 'Karl',
 '56': 'Ben',
 '57': 'Karl 2',
 '58': 'Karl Hampton',
 '59': 'Abby',
 '60': 'Kristi',
 '61': 'Abby',
 '62'

### Exercise

First we will get transactions for a specific user - as you can see in the documentation, you can use the `get_transactions` endpoint here and eppend the user_id of the user you want to retrieve data from

Get transaction data for our first user (check the endpoint in the documentation, and don't forget the API key)

In [415]:
user_id = 45

url_endpoint = f"{BASE_URL}/api/get_transactions/{user_id}"

r = requests.get(url=url_endpoint, headers=api_key)

r.status_code, r.content

(200, b'[{"amount":1000000.0,"type":"CREDIT"}]\n')

As you can see in the documentation, you can add an optional parameter to your request. With APIs you usually add a `?` followed by the optional arguments you want to add:

In [448]:
user_id = 1
url_endpoint = f"{BASE_URL}/api/get_transactions/{user_id}"

r = requests.get(url=url_endpoint, headers=api_key)
r.json()

[{'amount': 100.0, 'type': 'DEBIT'},
 {'amount': 99.9, 'type': 'DEBIT'},
 {'amount': 1000.0, 'type': 'CREDIT'},
 {'amount': 1000.0, 'type': 'CREDIT'},
 {'amount': 1800.1, 'type': 'DEBIT'},
 {'amount': 1000.0, 'type': 'CREDIT'},
 {'amount': 1123581321345.55, 'type': 'CREDIT'},
 {'amount': 1123581321345.55, 'type': 'DEBIT'}]

This can quickly become harder to work with as you are adding more optional parameters. `requests` provides a better way to add such parameters.

### Exercise

Instead of using `?` followed by the parameters, create a new dictionary with the arguments you want to use and their value, then pass this dictionary to `get()` as keyword argument `params`.

Syntax:

```
requests.get(your_url, params=your_parameters, headers=your_headers)
```

In [417]:
url_endpoint = f"{BASE_URL}/api/users"

r = requests.get(url=url_endpoint, headers=api_key)

r.status_code

200

Our web API also supports some `POST` operations where you can add data to our database. The first one we will see here is `add_user` that allows you to add a new user.

With `requests` we can simply use the `.post()` method and pass the data we want to send as a `data` parameter.

Replace the name in the dictionary below by your own name. This defines the data about a user that we want to send to the API.

In [418]:
# Replace by a name of your choice
user = {"name": "buckoii"}

In [419]:
# Here we have a new argument, data, that allows us to post data to the API
url_endpoint = f"{BASE_URL}/api/add_user"

r = requests.post(url_endpoint, data=user, headers=api_key)

Let's check the status code:

In [420]:
r.status_code

200

`200` means the operation was successful! If we call the get_users endpoint again (like we did at the very beginning) we should see our new user.

Get the updated list of users to find your new user's ID:

In [421]:
url_endpoint = f"{BASE_URL}/api/users"

r = requests.get(url=url_endpoint, headers=api_key)

r.status_code, r.content

(200,
 b'{"1":"Caroline","2":"Marium","3":"John","4":"John","5":"John","6":"Vic_Flores","7":"Payton","8":"Jess","9":"Sam","10":"Luning","11":"Karl","12":"Johan","13":"username","14":"test","15":"Bethan","16":"Linying","17":"Linying","18":"Vic_Flores","19":"\\ud83d\\ude00","20":"\\");DROP TABLE * ;","21":"D&G","22":"Ben S","23":"Ben S","24":"Michael","25":"John","26":"Emma","27":"Emma","28":"Jasmin","29":"Leo","30":"Matt","31":"Rasool","32":"John","33":"John","34":"John","35":"Amy","36":"Jeremy","37":"Xiong","38":"John","39":"John","40":"John","41":"John","42":"John","43":"Jeremy","44":"John","45":"Maxi Madcock","46":"Tilly Tats","47":"Raine The Paine","48":"John","49":"John","50":"John","51":"John","52":"Jon","53":"Zak","54":"John","55":"Karl","56":"Ben","57":"Karl 2","58":"Karl Hampton","59":"Abby","60":"Kristi","61":"Abby","62":"Benjo","63":"JonB","64":"ZAK","65":"Samantha","66":"John","67":"John","68":"Oli","69":"Oli","70":"Barry","71":"Jennie","72":"Rik","73":"Fahis","74":"Bobandy"

Finally, our web API allows us to add transactions by sending data serialised in json through a `POST` request to the same endpoint we've seen before. With `requests` we can simply use the `.post()` method and pass the data we want to send as a `data` parameter.

In [422]:
# Same endpoint as before, but here we'll use a POST request
user_id = 144 
url_endpoint = f"{BASE_URL}/api/add_transaction/{user_id}"

# The advantage of using our own API is that we can credit accounts as we want :)
transaction_to_add = {"type": "CREDIT", "amount": 1500000}
# transaction_to_add = {"type": "DEBIT", "amount": 200000}

r = requests.post(url_endpoint, data=transaction_to_add, headers=api_key)

In [423]:
# We verify the status code
type(r.status_code), r.content

(int, b'')

### Exercise

Let's check that our transaction was added; call the endpoint to `get_transactions` for the account you credited.

- use a `.get()` request with the necessary `headers` and the same `user_id` as used above

In [452]:
user_id = 3
url_endpoint = f"{BASE_URL}/api/get_transactions/{user_id}"

r = requests.get(url=url_endpoint, headers=api_key)
r.json()

[{'amount': 1000.0, 'type': 'CREDIT'}, {'amount': 1000.0, 'type': 'DEBIT'}]

### [Optional] A more advanced exercise

By now you should have all the tools you need to write your own programmes that leverage the power of APIs and automate tasks.

For instance here you should be able to write a simple programme that compute the balance for a given user.

For this you will need to:
- retrieve all CREDIT transactions for a user
- compute the sum of credits
- retrieve all DEBIT transactions for a user
- compute the sum of debit
- compute the balance

In [464]:
# Add your code below to compute the balance for a given user!


import numpy as np

# extension activity - replace transaction instances with transaction classs

# class BankTransaction:
#     def __init__(self, amount, tag):
#         self.amount = amount
#         self.tag = tag

        
class AccountLoader():
    
    def __init__(self):
        self.base_url = "https://europe-west2-kate-dev.cloudfunctions.net/banking"
        with open('data/api_key.json','r') as file:
            self.api_key = json.load(file)
        
        
    def get_credit_transactions(self, user_id):
        self.user_id = user_id
        credit_url = f"{self.base_url}/api/get_transactions/{self.user_id}?type=CREDIT"
        r = requests.get(credit_url, headers=self.api_key)
        
        if not r.status_code == 200:
            print('bad status code:')
            return {'status_code':r.status_code,'content': r.content}
        
        self.credit_transactions = [t['amount'] for t in json.loads(r.text)]
       
    
    def get_debit_transactions(self, user_id):
        self.user_id = user_id
        credit_url = f"{self.base_url}/api/get_transactions/{self.user_id}?type=DEBIT"
        r = requests.get(credit_url, headers=self.api_key)
        
        if not r.status_code == 200:
            print('bad status code:')
            return {'status_code':r.status_code,'content': r.content}
        
        self.debit_transactions = [t['amount'] for t in r.json()]
    
    def calculate_balance(self, user_id):
        if not hasattr(self, 'credit_transactions'):
            self.get_credit_transactions(user_id)
        
        if not hasattr(self, 'debit_transactions'):
            self.get_debit_transactions(user_id)
        
        self.balance = np.subtract(np.sum(self.credit_transactions), np.sum(self.debit_transactions))
    
    
def test_calculate_balance():
    cases = [{'user_id':1, 'balance':1000},
             {'user_id':3, 'balance':0} 
            ]
    
    for case in cases:
        loader = AccountLoader()
        loader.calculate_balance(case['user_id'])
        result = loader.balance

        assert case['balance'] == result
        
    print('\ncalculate_balance - test passed')

user_id = 144
    
loader = AccountLoader()
loader.calculate_balance(user_id)
print('my balance:', loader.balance)

test_calculate_balance()

my balance: 9700000.0

calculate_balance - test passed
